### PR TITLE
external_deps: better compression for archives

### PR DESCRIPTION
**Pull request by illwieckz**
If we're going to make new deps packages because of #712, it may be time to also implement this.

macOS doesn't provide `xz` binary but its `tar` binary is built with xz support.
Linux distros support `xz` since a very long time.

For Windows, we may use 7z for better zip, though maybe CMake knows how to extract `tar.xz` on Windows as well?

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -222,10 +222,10 @@ elseif (WIN32)
         set(DEPS_DIR ${EXTERNAL_DEPS_DIR}/mingw32-${DEPS_VERSION})
     endif()
 elseif (APPLE)
-    set(DEPS_EXT ".tar.bz2")
+    set(DEPS_EXT ".tar.xz")
```
Comment (DolceTriade) [inline CMakeLists.txt:225]: Pretty sure OSX doesn't ship with XZ by default.

**Comment by DolceTriade** [comment]
Can we get numbers to compare xz vs bz2 vs zip on each platform?

File: CMakeLists.txt (hunk 1 above)
Comment (illwieckz) [reply CMakeLists.txt:225]: No `xz` binary but It looks like tar has lzma support. Example on Mojave (which is older than officially supported macOS releases):

```
$ otool -L /usr/bin/tar
/usr/bin/tar:
	/usr/lib/libarchive.2.dylib (compatibility version 9.0.0, current version 9.2.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.250.1)
	
$ otool -L /usr/lib/libarchive.2.dylib
/usr/lib/libarchive.2.dylib:
	/usr/lib/libarchive.2.dylib (compatibility version 9.0.0, current version 9.2.0)
	/usr/lib/libbz2.1.0.dylib (compatibility version 1.0.0, current version 1.0.5)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libxml2.2.dylib (compatibility version 10.0.0, current version 10.9.0)
	/usr/lib/liblzma.5.dylib (compatibility version 6.0.0, current version 6.3.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.250.1)
```

```
$ touch a
$ tar cJvf a.tar.xz a
a a
$ file a.tar.xz 
a.tar.xz: XZ compressed data
$ tar tf a.tar.xz 
a
```

**Comment by illwieckz** [comment]
Some compression comparison.

```
uncompressed

  170 580 507	linux64-6
  280 582 752	macosx64-6
  267 267 862	mingw32-6
  266 308 219	mingw64-6
  264 082 928	msvc32-6
  262 642 150	msvc64-6
1 511 464 418	total

bz2, zip (current options)

  50 178 734	linux64-6.tar.bz2
  82 156 952	macosx64-6.tar.bz2
  95 501 401	mingw32-6.zip
  95 170 121	mingw64-6.zip
  95 122 773	msvc32-6.zip
  94 680 294	msvc64-6.zip
 512 810 275	total

xz -9, zip -9

  30 708 176	linux64-6-z9.tar.xz
  38 297 140	macosx64-6-z9.tar.xz
  95 022 755	mingw32-6-z9.zip
  94 704 494	mingw64-6-z9.zip
  94 668 238	msvc32-6-z9.zip
  94 236 132	msvc64-6-z9.zip
 447 636 935	total

xz -9, 7z -tzip -mx=9

  30 708 176	linux64-6-z9.tar.xz
  38 297 140	macosx64-6-z9.tar.xz
  92 030 399	mingw32-6-z9.zip
  91 706 501	mingw64-6-z9.zip
  91 686 471	msvc32-6-z9.zip
  91 261 031	msvc64-6-z9.zip
 435 689 718	total

xz -9, 7z -tzip -mx=9 -mfb=258 -mpass=15
  30 708 176	linux64-6-z9.tar.xz
  38 297 140	macosx64-6-z9.tar.xz
  92 003 242	mingw32-6-z9.zip
  91 680 318	mingw64-6-z9.zip
  91 662 018	msvc32-6-z9.zip
  91 236 521	msvc64-6-z9.zip
 435 587 415	total

xz -9, some crazy zip optimizations

  30 708 176	linux64-6-z9.tar.xz
  38 297 140	macosx64-6-z9.tar.xz
  91 819 459	mingw32-6-z9.zip
  91 498 545	mingw64-6-z9.zip
  91 476 300	msvc32-6-z9.zip
  91 053 419	msvc64-6-z9.zip
 434 853 039	total

xz -9

  30 708 176	linux64-6-z9.tar.xz
  38 297 140	macosx64-6-z9.tar.xz
  35 382 764	mingw32-6-z9.tar.xz
  35 106 832	mingw64-6-z9.tar.xz
  35 246 544	msvc32-6-z9.tar.xz
  34 929 384	msvc64-6-z9.tar.xz
 209 670 840	total
 ```

**Comment by illwieckz** [comment]
I'm trying `tar.xz` for every platform to see  how it behaves. The `tar.xz` is expected to be extracted by cmake so maybe cmake can extract it on every plaform.

If I'm right linux, mingw, and msvc packages are done on Linux so the script is expected to find a tar with xz support, and it looks like macos has tar with xz support. So if cmake support extracting `tar.xz` on every system and we can produce `tar.xz` for every system, we may want to do that (Appveyor hasn't started to build yet so I donet know yet).

**Comment by illwieckz** [comment]
So, it looks like `tar.xz` works everywhere, it save file size by 59%, from 512M to 209M.

**Comment by illwieckz** [comment]
On a side topic, I suggested in #561 to switch to this kind of naming scheme: `<system>-<arch>-<compiler>_<version>.<ext>`, so basically we would have:

```
linux-amd64-default_7.tar.xz
macos-amd64-default_7.tar.xz
windows-amd64-mingw_7.tar.xz
windows-amd64-msvc_7.tar.xz
windows-i686-mingw_7.tar.xz
windows-i686-msvc_7.tar.xz
```

I can do it while I'm at it. What do you think about it?

**Comment by DolceTriade** [comment]
It works with MSVC too? Does CMake extract it?

File: external_deps/build.sh [hunk 2]
```diff
@@ -623,13 +623,10 @@ build_install() {
...
 	cd "${WORK_DIR}"
 	case "${PLATFORM}" in
-	mingw*|msvc*)
-		rm -f "${PLATFORM}-${DEPS_VERSION}.zip"
-		zip -r "${PLATFORM}-${DEPS_VERSION}.zip" "${PLATFORM}-${DEPS_VERSION}"
-		;;
 	*)
-		rm -f "${PLATFORM}-${DEPS_VERSION}.tar.bz2"
-		tar cvjf "${PLATFORM}-${DEPS_VERSION}.tar.bz2" "${PLATFORM}-${DEPS_VERSION}"
+		export XZ_OPT='-9'
```
Comment (slipher) [inline external_deps/build.sh:627]: Would rather see `XZ_OPT=-9 tar cvJf`... than set a variable that persists for the rest of the program.

**Comment by slipher** [comment]
> On a side topic, I suggested in #561 to switch to this kind of naming scheme: `<system>-<arch>-<compiler>_<version>.<ext>`

Sounds good

Do we know if .tar.xz works on Windows yet?

**Comment by DolceTriade** [comment]
This LGTM assuming it works on Windows.

**Comment by illwieckz** [comment]
The CI worked on Windows. @slipher can you try this branch yourself on Windows?

**Comment by DolceTriade** [comment]
```
-- Downloading dependencies from 'https://dl.unvanquished.net/deps/msvc64-6.tar.xz'
-- [download 0% complete]
-- [download 1% complete]
-- [download 2% complete]
-- [download 3% complete]
-- [download 4% complete]
-- [download 5% complete]
-- [download 6% complete]
-- [download 7% complete]
-- [download 8% complete]
-- [download 9% complete]
-- [download 10% complete]
-- [download 11% complete]
-- [download 12% complete]
-- [download 13% complete]
-- [download 14% complete]
-- [download 15% complete]
-- [download 16% complete]
-- [download 17% complete]
-- [download 18% complete]
-- [download 19% complete]
-- [download 20% complete]
-- [download 21% complete]
-- [download 22% complete]
-- [download 23% complete]
-- [download 24% complete]
-- [download 25% complete]
-- [download 26% complete]
-- [download 27% complete]
-- [download 28% complete]
-- [download 29% complete]
-- [download 30% complete]
-- [download 31% complete]
-- [download 32% complete]
-- [download 33% complete]
-- [download 34% complete]
-- [download 35% complete]
-- [download 36% complete]
-- [download 37% complete]
-- [download 38% complete]
-- [download 39% complete]
-- [download 40% complete]
-- [download 41% complete]
-- [download 42% complete]
-- [download 43% complete]
-- [download 44% complete]
-- [download 45% complete]
-- [download 46% complete]
-- [download 47% complete]
-- [download 48% complete]
-- [download 49% complete]
-- [download 50% complete]
-- [download 51% complete]
-- [download 52% complete]
-- [download 53% complete]
-- [download 54% complete]
-- [download 55% complete]
-- [download 56% complete]
-- [download 57% complete]
-- [download 58% complete]
-- [download 59% complete]
-- [download 60% complete]
-- [download 61% complete]
-- [download 62% complete]
-- [download 63% complete]
-- [download 64% complete]
-- [download 65% complete]
-- [download 66% complete]
-- [download 68% complete]
-- [download 69% complete]
-- [download 71% complete]
-- [download 72% complete]
-- [download 76% complete]
-- [download 77% complete]
-- [download 79% complete]
-- [download 81% complete]
-- [download 85% complete]
-- [download 86% complete]
-- [download 88% complete]
-- [download 91% complete]
-- [download 92% complete]
-- [download 95% complete]
-- [download 99% complete]
-- [download 100% complete]
-- Download completed successfully
-- Found OpenGL: opengl32   
-- Found ZLIB: C:/projects/daemon/external_deps/msvc64-6/lib/libz.lib (found version "1.2.12") 
-- Found Nettle: C:/projects/daemon/external_deps/msvc64-6/lib/libnettle.lib  
-- Found GMP: C:/projects/daemon/external_deps/msvc64-6/lib/libgmp.lib
-- Found CURL: C:/projects/daemon/external_deps/msvc64-6/lib/libcurl.lib (found version "7.[83](https://ci.appveyor.com/project/DolceTriade/daemon/builds/44869713/job/y4f4pjbdpficdkp5#L83).1")  
-- Looking for pthread.h
-- Looking for pthread.h - not found
-- Found Threads: TRUE  
-- Found SDL2: C:/projects/daemon/external_deps/msvc64-6/lib/SDL2main.lib;C:/projects/daemon/external_deps/msvc64-6/lib/SDL2.lib  
-- Found Ogg: C:/projects/daemon/external_deps/msvc64-6/lib/libogg.lib  
-- Found Vorbis: C:/projects/daemon/external_deps/msvc64-6/lib/libvorbis.lib  
-- Found Opus: C:/projects/daemon/external_deps/msvc64-6/lib/libopusfile.lib  
-- Found WebP: C:/projects/daemon/external_deps/msvc64-6/lib/libwebp.lib  
-- Performing Test _WEBP_COMPILATION_TEST
-- Performing Test _WEBP_COMPILATION_TEST - Success
-- Found JPEG: C:/projects/daemon/external_deps/msvc64-6/lib/libjpeg.lib (found version "80") 
-- Found PNG: C:/projects/daemon/external_deps/msvc64-6/lib/libpng.lib (found version "1.6.37") 
-- Found Freetype: C:/projects/daemon/external_deps/msvc64-6/lib/libfreetype.lib (found version "2.12.1") 
-- Found GLEW: C:/projects/daemon/external_deps/msvc64-6/lib/libglew32.lib  
-- Found OpenAL: C:/projects/daemon/external_deps/msvc64-6/lib/libOpenAL32.lib
```seems good

**Comment by illwieckz** [comment]
So it looks to be working with `tar.xz` on every platform.

**Comment by illwieckz** [comment]
Commit moved to #712.

**Pull request closed without merging**